### PR TITLE
arm: linker.ld: move bss section to ram end

### PR DESCRIPTION
**Pull request by lgl88911**
fix bss section be copy to binary

Signed-off-by: Frank Li <lgl88911@163.com>

### CODE REVIEW

File: include/arch/arm/aarch32/cortex_m/scripts/linker.ld [hunk 1]
```diff
@@ -342,42 +377,6 @@ SECTIONS
...
-    __data_rom_start = LOADADDR(_DATA_SECTION_NAME);
-
-#include <linker/common-ram.ld>
-#include <linker/priv_stacks.ld>
-#include <linker/kobject.ld>
-
-#include <linker/priv_stacks-noinit.ld>
-#include <linker/cplusplus-ram.ld>
-
     __data_ram_end = .;
```
Comment (stephanosio) [inline include/arch/arm/aarch32/cortex_m/scripts/linker.ld]: `__data_ram_end` must be moved too.

**Review by stephanosio — CHANGES_REQUESTED** [review]
Please provide an explanation as to why this is necessary in the commit message.

**Comment by ioannisg** [comment]
> Please provide an explanation as to why this is necessary in the commit message.

agreed. What problem are you trying to address @lgl88911 ?

**Comment by lgl88911** [comment]
@stephanosio  @ioannisg Thanks for your review
When I hope load code to ram and run it, I compiling with non-xip(CONFIG_XIP=n), all sections will be placed in the ram. If NOBITS section is not placed at the end of the ram, these sections will be placed to binary file. eg: bss, noinit
You can reduce by using mimxrt1050_evk, follow below step:
1. Add **CONFIG_XIP=n** to mimxrt1050_evk_defconfig
2. build it with any sample
3. check zephyr.bin size
4. add large global variables to same sample and use these global variables to make sure it is not removed for optimization
5. build it again, you will find the size of zephyr.bin has grown

File: include/arch/arm/aarch32/cortex_m/scripts/linker.ld (hunk 1 above)
Comment (lgl88911) [reply include/arch/arm/aarch32/cortex_m/scripts/linker.ld]: Sorry for lost __data_ram_end, I move it too, Thansk !

**Comment by stephanosio** [comment]
While I can confirm that behaviour, something is wrong here ...

```
/opt/zephyr-sdk/arm-zephyr-eabi/bin/arm-zephyr-eabi-objcopy -S --gap-fill 0xff --output-target=binary --remove-section=.comment --remove-section=COMMON --remove-section=.eh_frame zephyr.elf zephyr.bin
```

CONFIG_XIP=y
```
-rwxrwxr-x 1 stephanos stephanos 21464 Jan  8 01:24 zephyr/zephyr.bin
```

CONFIG_XIP=n
```
-rwxrwxr-x 1 stephanos stephanos 90676 Jan  8 01:25 zephyr/zephyr.bin
```

https://gist.github.com/stephanosio/a7f5cd34c7480bacb5254d4760b54805

The `bss` and `noinit` sections are marked `NOBITS` and the objcopy should not be putting them in the bin file. Maybe I am missing something .. I will look into this tomorrow.

**Comment by stephanosio** [comment]
> If NOBITS section is not placed at the end of the ram, these sections will be placed to binary file. eg: bss, noinit

To be exact, it is not that the `NOBITS` sections are being "placed" into the output binary file; but rather, the gap between the first `NOBITS` section and the next `PROGBITS` section must be filled before placing the data `PROGBITS` sections (note that LMA equals VMA in non-XIP images).

**Review by stephanosio — COMMENTED** [review]
I am not against this change because it does help reduce the output 'bin' file size, which can be useful if you must use the 'bin' image (e.g. booting from a 'bin' file in an SD card).

However, this seems to be breaking some existing platforms and they will need to be fixed before this PR can be approved.

**Comment by lgl88911** [comment]
> I am not against this change because it does help reduce the output 'bin' file size, which can be useful if you must use the 'bin' image (e.g. booting from a 'bin' file in an SD card).
> 
> However, this seems to be breaking some existing platforms and they will need to be fixed before this PR can be approved.

Thanks for your check.

Yes, I hope load bin from SD card to sdram, Decreasing bin size can help improve download and load speeds。

What kind of problems this change will affect those existing platforms？

**Comment by lgl88911** [comment]
> > I am not against this change because it does help reduce the output 'bin' file size, which can be useful if you must use the 'bin' image (e.g. booting from a 'bin' file in an SD card).
> > However, this seems to be breaking some existing platforms and they will need to be fixed before this PR can be approved.
> 
> Thanks for your check.
> 
> Yes, I hope load bin from SD card to sdram, Decreasing bin size can help improve download and load speeds。
> 
> What kind of problems this change will affect those existing platforms？

I know "EXISTING PLATFORMS" problem is :
some object address in zephyr_prebuilt.elf diff to zephyr.elf.  I'm trying to find the reason

**Comment by lgl88911** [comment]
@stephanosio Shippable check pass, Could you help review again,Thanks

File: include/arch/arm/aarch32/cortex_m/scripts/linker.ld [hunk 2]
```diff
@@ -296,9 +296,8 @@ SECTIONS
 
 	_app_smem_size = _app_smem_end - _app_smem_start;
 	_app_smem_rom_start = LOADADDR(_APP_SMEM_SECTION_NAME);
-#endif  /* CONFIG_USERSPACE */
 
-    SECTION_DATA_PROLOGUE(_BSS_SECTION_NAME,(NOLOAD),)
+   SECTION_DATA_PROLOGUE(_BSS_SECTION_NAME,(NOLOAD),)
```
Comment (stephanosio) [inline include/arch/arm/aarch32/cortex_m/scripts/linker.ld]: This whitespace change should not be present; that is, unless you are willing to clean up the entire file while you are at it.

File: include/arch/arm/aarch32/cortex_m/scripts/linker.ld [hunk 3]
```diff
@@ -379,8 +380,55 @@ SECTIONS
 #include <linker/priv_stacks-noinit.ld>
 #include <linker/cplusplus-ram.ld>
 
-    __data_ram_end = .;
+__data_ram_end = .;
```
Comment (stephanosio) [inline include/arch/arm/aarch32/cortex_m/scripts/linker.ld]: same here

File: include/arch/arm/aarch32/cortex_m/scripts/linker.ld [hunk 4]
```diff
@@ -379,8 +380,55 @@ SECTIONS
...
+/* Located in generated directory. This file is populated by the
+ * zephyr_linker_sources() Cmake function.
+ */
+#include <snippets-noinit.ld>
+#ifdef CONFIG_SOC_NOINIT_LD
+#include <soc-noinit.ld>
+#endif
+
+        } GROUP_LINK_IN(RAMABLE_REGION)
+#endif /*CONFIG_USERSPACE*/
```
Comment (stephanosio) [inline include/arch/arm/aarch32/cortex_m/scripts/linker.ld]: To be consistent with others: `/* CONFIG_USERSPACE */`

**Review by stephanosio — COMMENTED** [review]
So, in summary, we cannot have the `bss` section as the last RAM section when `USERSPACE` is enabled; I wonder why that is.

IMO, we should figure out why the userspace tests fault when the `bss` section is placed as the last section in the RAM and provide a more fundamental fix, unless there is a technical reason as to why that is not possible and/or feasible.

I don't have time to look into it right now; if it does not get resolved in time, I will eventually come back to it.

**Comment by lgl88911** [comment]
> So, in summary, we cannot have the `bss` section as the last RAM section when `USERSPACE` is enabled; I wonder why that is.
> 
> IMO, we should figure out why the userspace tests fault when the `bss` section is placed as the last section in the RAM and provide a more fundamental fix, unless there is a technical reason as to why that is not possible and/or feasible.
> 
> I don't have time to look into it right now; if it does not get resolved in time, I will eventually come back to it.

When bss moves to the end, the bss section will be after the kobject_data section.
Uninitialized kernel objects are placed in bss.
When userspace is enabled, there is no data in kobject_data during the first stage of linking(prelink), so the kernel object address is  small.
Before the second stage of linking(finial), the kernel objects in prelink are scanned and their addresses are placed in kobject_data.
At this time, kobject_data becomes larger, bss be shifted backwards , cause the kernel object address in the final link result will become larger, which will be inconsistent with that in kobject_data.
If the position of bss is before kobject_data, there will not be this problem. 
The appropriate method may be to reserve space for kobject_data during prelink, but the related process is relatively unfamiliar to me, so it is difficult to modify.
The current modification can at least solve the non-userspace situation, which is enough for my application.

File: include/arch/arm/aarch32/cortex_m/scripts/linker.ld (hunk 4 above)
Comment (lgl88911) [reply include/arch/arm/aarch32/cortex_m/scripts/linker.ld]: Can you give some suggestions, is there a formal way to distinguish the two `/ * CONFIG_USERSPACE * / `  Thanks

**Comment by andrewboie** [comment]
> The appropriate method may be to reserve space for kobject_data during prelink, but the related process is relatively unfamiliar to me, so it is difficult to modify.

The relevant linker snippet is include/linker/kobject.ld.

I don't have a good suggestion to fix this at the moment. The problem is that the size of this area is a linear function of the number of kernel objects in the binary, there isn't a good way to predict how large it is without going through the full process of building.

**Comment by lgl88911** [comment]
> > The appropriate method may be to reserve space for kobject_data during prelink, but the related process is relatively unfamiliar to me, so it is difficult to modify.
> 
> The relevant linker snippet is include/linker/kobject.ld.
> 
> I don't have a good suggestion to fix this at the moment. The problem is that the size of this area is a linear function of the number of kernel objects in the binary, there isn't a good way to predict how large it is without going through the full process of building.

Indeed, the inability to predict kernel object data is a problem. Maybe reserve enough space, or add a third link, use the final elf to generate kobject_hash.c again, and relink the new elf.
But both methods are "not pretty", so my idea is to make this scheme work on non-userspace first.

File: include/arch/arm/aarch32/cortex_m/scripts/linker.ld (hunk 4 above)
Comment (stephanosio) [reply include/arch/arm/aarch32/cortex_m/scripts/linker.ld]: I meant the space after and before `*`.

`/* CONFIG_USERSPACE */`
`/*CONFIG_USERSPACE*/`

**Comment by stephanosio** [comment]
> Indeed, the inability to predict kernel object data is a problem. Maybe reserve enough space, or add a third link, use the final elf to generate kobject_hash.c again, and relink the new elf.
> But both methods are "not pretty", so my idea is to make this scheme work on non-userspace first.

Thanks for the explanation. It looks like that will require a major overhaul, which should probably be a separate PR.

**Review by stephanosio — APPROVED** [review]
Approving for now, noting that it is not feasible for userspace to have the bss as the last section in the RAM.

File: include/arch/arm/aarch32/cortex_m/scripts/linker.ld (hunk 4 above)
Comment (stephanosio) [reply include/arch/arm/aarch32/cortex_m/scripts/linker.ld]: Also, although not strictly a rule, it is common to do add `!` for `ifndef`s, e.g. `/* !CONFIG_USERSPACE */` to distinguish from `ifdef`s.

Comment (lgl88911) [reply include/arch/arm/aarch32/cortex_m/scripts/linker.ld]: Thanks for your explain, I fixed it

**Comment by lgl88911** [comment]
@galak @MaureenHelm @andrewboie @ioannisg This PR has been pending for a long time, can you help review if need to make other changes?

**Comment by carlescufi** [comment]
> @galak @MaureenHelm @andrewboie @ioannisg This PR has been pending for a long time, can you help review if need to make other changes?

Yep, this should be followed-up. @lgl88911 if you get no response before tomorrow I can suggest you add it to the dev-review meeting and [join](https://github.com/zephyrproject-rtos/zephyr/wiki/Zephyr-Committee-and-Working-Group-Meetings#zephyr-dev-meeting) if you are able to.

**Review by tyrantlizard77 — APPROVED** [review]
Fine with this for now, but I'm going to open a GH enhancement to investigate this more deeply, such that if user mode is enabled on a non-XIP system the binary isn't inflated by NOBITS sections.

**Pull request merged**